api/app.py

```diff
@@ -10,67 +13,257 @@
 
-@app.route('/api/recipes/<int:recipe_id>', methods=['GET'])
-def get_recipe(recipe_id):
+def get_recipe_as_list():
     recipes = []
     for recipe in db.session.query(Recipe).all():
-        recipes.append(recipe.as_dict())
-    for recipe in recipes:
-        if recipe['id'] == recipe_id:
-            return jsonify(recipe)
-    return 'Recipe not found', 404
+        recipes.append(recipe_json(recipe))
+    return recipes
+
+@app.route('/api/recipes', methods=['GET'])
+def get_recipes():
+    recipes = get_recipe_as_list()
+    return jsonify(recipes)
 
 
 @app.route('/api/recipes', methods=['POST'])
 def create_recipe():
-    new_recipe = {
-        'id': len(recipes) + 1,
-        'name': request.json.get('name'),
-        'duration': request.json.get('duration'),
-        'pictures': request.json.get('pictures'),
-        'instructions': request.json.get('instructions'),
-        'categories': request.json.get('categories'),
-        'ingredients': request.json.get('ingredients'),
-    }
-    recipes.append(new_recipe)
-    return jsonify(new_recipe)
+    data = request.get_json()
+    if data is None:
+        return jsonify({'message': 'No input data provided'}), 400
+
+    name = data.get('name')
+    if not name  or not isinstance(name, str) or name.strip() == "":
+        return jsonify({'message': 'Name is required'}), 400
+
+    duration = data.get('duration')
+    if not duration or not isinstance(duration, str) or duration.strip() == "":
+        return jsonify({'error': 'Duration must be a non-empty string'}), 400
+
+    pictures = data.get('pictures')
+    if not pictures or not isinstance(pictures, list) or not all(isinstance(pic, str) for pic in pictures):
+        return jsonify({'message': 'Pictures are required as a list of valid URLs'}), 400
+
+    instructions = data.get('instructions')
+    if not instructions or not isinstance(instructions, str) or instructions.strip() == "":
+        return jsonify({'message': 'Instructions are required'}), 400
+
+    categories = []
+    categories_data = data.get('categories', [])
+    for category in categories_data:
+        if isinstance(category, str):
+            name = category
+            color = '#808080'
+        elif isinstance(category, dict):
+            name = category['name']
+            color = category.get('color', '#808080')
+        else:
+            return jsonify({'message': 'Invalid format'}), 400
+        if not name:
+            jsonify({'message': 'No name provided'}), 400
+        cat = Category.query.filter_by(name=name).first()
+        if not cat:
+            cat = Category(name=name, color=color)
+            db.session.add(cat)
+        categories.append(cat)
+
+    ingredient_regex = re.compile(r'^(?P<quantity>\d+)(?P<unit>[a-zA-Z]*) (?P<name>.+)$')
+    ingredients = []
+    ingredients_data = data.get('ingredients', [])
+    for ingredient_str in ingredients_data:
+        match = ingredient_regex.match(ingredient_str.strip())
```

Comment:
We are going to receive both `categories` and `ingredients` as lists of dicts from frontend, so they'll have the format:
```
"categories": 
  [
    {
      "name": "Pie",
      "color": "#FFD700"
    },
    {
      "name": "Italian",
      "color": "#FF0000"
    }
  ]
```
```
"ingredients": [
      {
        "name": "sugar",
        "unit": "g",
        "quantity": 150
      },
      {
        "name": "eggs",
        "unit": null,
        "quantity": 3
      }
    ]
```
We don't need to separate the `ingredients`, they are already separated into the 3 values that we need (`name`, `unit`, and `quantity`).